### ISSUE TITLE
Allow the keyboard to write the password on screenlock after reboot

### ISSUE THREAD

**Issue by T-3B** [Julow/Unexpected-Keyboard#216]
On Android 11 (LineAgeOS), I cannot use your application to enter my password (needed to unlock my phone).

If I have only the Unexpected keyboard, no keyboard show up when clicking on the field. I needed a hardware keyboard (OTG) to unlock my phone.

If I have another keyboard (like com.touchtype.swiftkey.beta - Microsoft Swiftkey Beta), that other keyboard will show up for entering the password even if Unexpected is set to be default keyboard.

It seems to me that Unexpected Keyboard **does not have the right permission**.
I even have a message from my system when I select Unexpected Keyboard (only first time agter fresh installation) that I don't have with SwiftKey Beta:
![IMG_20221012_084740](https://user-images.githubusercontent.com/87860429/195272388-9385a1b7-51b6-4d87-ad8a-50dacc764f5c.jpg)
![IMG_20221012_084800](https://user-images.githubusercontent.com/87860429/195272391-004e4209-7b67-4b85-97cd-b658c211e787.jpg)
![IMG_20221012_084828](https://user-images.githubusercontent.com/87860429/195272393-2aff77a6-434c-4b1c-bde8-178fe9881074.jpg)
![IMG_20221012_084850](https://user-images.githubusercontent.com/87860429/195272395-9ecc2a1c-9e1a-4e0b-a60e-db58e625a7cc.jpg)

Please note that I have no other keyboard, I'm on a custom ROM and I removed the pre-installed keyboard.
I really love your keyboard, and would love to only use it (even for password on lock screen) !
Hope it can be easily implemented.

**Comment by parkerlreed** [comment]
User installed keyboards are unable to draw on lockscreen for security reasons.

**Comment by T-3B** [comment]
Hi !

Thanks for your answer.

But i'm sure user installed keyboards *can* write lockscreen passwords.
For example it is the case of "Microsoft SwiftKey Beta".

Also, Unexpected keyboard can write passwords in lockscreen, but needs to 
be allowed after each reboot. Thus if I only have Termux keyboard and 
reboot, I would need a hardware keyboard (via OTG) to unlock my phone and 
then re-allow Unexpected keyboard to write passwords on lockscreen (which 
is a bit annoying).

On the other hand, it seems that SwiftKey keyboard keeps the permission to 
write passwords on lockscreen, even after reboot.
I can ensure you that I *manually* installed/downloaded Swiftkey, so the 
problem is not there.

I wish the problem will be resolved, as it seems (to me) feasible.

Don't hesitate to ask any help (or screen shots/records)!

Thanks for your great app,
SMAGGHE Clément.

**Comment by Julow** [comment]
Thanks for the screenshots. Added in 25cefc5.

**Comment by T-3B** [comment]
Thanks.

I have a good and a bad news:
- the good one is that my system does not show anymore the message "Note: If you restart your phone and have a screen lock set […]" (I even uninstalled Unexpected Keyboard and then installed 1.19.0, to reconfigure it properly in the system. So apparently it should now work (and show up) just after a reboot for the screen lock password.
- the bad one: unfortunately the keyboard does not show after reboot when (trying) entering my password, but for another reason. Thus I needed to plug an hardware keyboard via OTG, and once unlocked my phone showed me this message:
![Screenshot_20221106-220741_Nova7](https://user-images.githubusercontent.com/87860429/200195778-0dd3ca25-3765-4368-b5e9-c4858b74a4d9.png)
So basically Unexpected Keyboard just crashed.
I wanted to test with another keyboard (SwiftKey Beta) to see if the app crashes again, and yes it crashes. Even if I set Swiftkey Beta as the default keyboard, then when I need to enter my password I select the Unexpected Keyboard as input, Unexpected won't show up.

Thanks for your work, and once again don't hesitate to ask screenshots and other stuffs!

**Comment by Julow** [comment]
Thanks for testing! What's your version of android ? Do you think you could give me the crash report ? (with `adb logcat`, the stack trace that mentions `juloo.keyboard2`)

**Comment by T-3B** [comment]
My phone is running under Android 11 (LineAgeOS but that should not matter).
It is also rooted with Magisk.

And sure, I'll send you the crash report soon. Stay tuned 😋

**Comment by darhma** [comment]
I had commented #238 confirming the problem with version 1.19.0 on android 12

**Comment by T-3B** [comment]
@Julow 
Here you can find the log!
Please see [here](https://litter.catbox.moe/klohdo.log) for a more complete log (*expires in 3 days), although it should loop beetween these errors:
```log
11-10 09:09:43.271  1534  2018 I SystemServerTimingAsync: InitThreadPoolExec:Update app-ops uidState in case package juloo.keyboard2 changed
11-10 09:09:43.271  1534  2018 D SystemServerInitThreadPool: Started executing Update app-ops uidState in case package juloo.keyboard2 changed
11-10 09:09:43.272  1534  2018 D SystemServerInitThreadPool: Finished executing Update app-ops uidState in case package juloo.keyboard2 changed
11-10 09:09:43.272  1534  2018 D SystemServerTimingAsync: InitThreadPoolExec:Update app-ops uidState in case package juloo.keyboard2 changed took to complete: 1ms
11-10 09:09:44.025  1534  1534 V StorageManagerService: Package juloo.keyboard2 does not have legacy storage
11-10 09:09:57.484  1534  1634 I ActivityManager: Start proc 3228:juloo.keyboard2/u0a453 for service {juloo.keyboard2/juloo.keyboard2.Keyboard2}
11-10 09:09:57.590  3228  3228 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/cache: mkdir failed: errno 126 (Required key not available)
11-10 09:09:57.618  3228  3228 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/shared_prefs: mkdir failed: errno 126 (Required key not available)
11-10 09:09:57.624  3228  3228 E AndroidRuntime: Process: juloo.keyboard2, PID: 3228
11-10 09:09:57.624  3228  3228 E AndroidRuntime: java.lang.RuntimeException: Unable to create service juloo.keyboard2.Keyboard2: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
11-10 09:09:57.624  3228  3228 E AndroidRuntime: 	at juloo.keyboard2.Keyboard2.onCreate(Keyboard2.java:52)
11-10 09:09:57.650  1534  2329 I ActivityManager: Process juloo.keyboard2 (pid 3228) has died: fg  IMPB
11-10 09:09:57.652  1534  2329 W ActivityManager: Scheduling restart of crashed service juloo.keyboard2/.Keyboard2 in 1000ms for connection
11-10 09:09:58.687  1534  1634 I ActivityManager: Start proc 3343:juloo.keyboard2/u0a453 for service {juloo.keyboard2/juloo.keyboard2.Keyboard2}
11-10 09:09:58.744  3343  3343 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/cache: mkdir failed: errno 126 (Required key not available)
11-10 09:09:58.770  3343  3343 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/shared_prefs: mkdir failed: errno 126 (Required key not available)
11-10 09:09:58.772  3343  3343 E AndroidRuntime: Process: juloo.keyboard2, PID: 3343
11-10 09:09:58.772  3343  3343 E AndroidRuntime: java.lang.RuntimeException: Unable to create service juloo.keyboard2.Keyboard2: java.lang.IllegalStateException: SharedPreferences in credential encrypted storage are not available until after user is unlocked
11-10 09:09:58.772  3343  3343 E AndroidRuntime: 	at juloo.keyboard2.Keyboard2.onCreate(Keyboard2.java:52)
11-10 09:09:58.797  1534  2315 I ActivityManager: Process juloo.keyboard2 (pid 3343) has died: fg  IMPB
11-10 09:09:58.797  1534  2315 W ActivityManager: Scheduling restart of crashed service juloo.keyboard2/.Keyboard2 in 1800000ms for connection
11-10 09:10:02.663  1534  1634 I ActivityManager: Start proc 3412:juloo.keyboard2/u0a453 for service {juloo.keyboard2/juloo.keyboard2.Keyboard2}
11-10 09:10:02.695  3412  3412 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/cache: mkdir failed: errno 126 (Required key not available)
11-10 09:10:02.706  3412  3412 W ContextImpl: Failed to ensure /data/user/0/juloo.keyboard2/shared_prefs: mkdir failed: errno 126 (Required key not available)
```



***TL;DR*** I suspect the problem is caused by `SharedPreferences in credential encrypted storage are not available until after user is unlocked`.

**Comment by Julow** [comment]
Thanks for the crash logs. It should be fixed in f1ce6ab.

I'd appreciate if you could try the fixed version. I'll otherwise release soon.
You can find a debug version here: https://github.com/Julow/Unexpected-Keyboard/actions/runs/3445400199

**Comment by T-3B** [comment]
Yep, works as expected! :partying_face:

This issue can be closed (and surely #238 too).

Thank you for your hard work, appreciate it.

**Comment by Julow** [comment]
Thanks :)
